llama.cpp

```diff
@@ -11322,30 +11372,13 @@ static int llama_decode_internal(
             // no output
             res  = nullptr;
             embd = nullptr;
-        } else if (!hparams.causal_attn) {
-            res = nullptr; // do not extract logits for embedding models such as BERT
-
-            // token or sequence embeddings
-            embd = gf->nodes[gf->n_nodes - 1];
-
-            GGML_ASSERT(strcmp(embd->name, "result_embd") == 0 || strcmp(embd->name, "result_embd_pooled") == 0);
         } else if (cparams.embeddings) {
-            // the embeddings could be in the second to last tensor, or any of the previous tensors
-            int i_embd = gf->n_nodes - 2;
-            for (int i = 3; strcmp(embd->name, "result_norm") != 0; ++i) {
-                i_embd = gf->n_nodes - i;
-                if (i_embd < 0) { break; }
-                embd = gf->nodes[i_embd];
-            }
-            GGML_ASSERT(i_embd >= 0 && "missing result_norm tensor");
-
-            // TODO: use a per-batch flag to know when to skip logits while keeping embeddings
-            if (!cparams.causal_attn) {
-                res = nullptr; // do not extract logits when not needed
-                // skip computing logits
-                // TODO: is this safe?
-                gf->n_nodes = i_embd + 1;
+            res = nullptr; // do not extract logits for embedding case
+            embd = gf->nodes[gf->n_nodes - 1];
+            if (strcmp(embd->name, "result_embd_pooled") != 0) {
+                embd = gf->nodes[gf->n_nodes - 2];
             }
+            GGML_ASSERT(strcmp(embd->name, "result_embd_pooled") == 0 && "missing embeddings tensor");
         } else {
             embd = nullptr; // do not extract embeddings when not needed
             GGML_ASSERT(strcmp(res->name, "result_output") == 0 && "missing result_output tensor");
```

Comment:
There are places that need to know when embeddings or logits will be output, like `llama_output_reserve`

https://github.com/ggerganov/llama.cpp/blob/cd93a28cb1446319af5e2f4b416174c3a8e43546/llama.cpp#L11064-L11065

This will need to be updated to reflect exactly how this affects what happens later in this function near the comments `// extract logits` and `// extract embeddings`.

Reply:
So can we get away with saying you're either getting logits or embeddings but never both, and that behavior is exclusively controlled by `cparams.embeddings`? In that case we could just have
```c++
const bool has_logits = !cparams.embeddings;
const bool has_embd   =  cparams.embeddings;
```

Reply:
Yeah, I can't really think of a use-case where both would be needed at the same time. ***Except*** *maybe* for a `server` serving both completions and embeddings out of the same model. So that's something to consider.

Reply:
Agreed, but for a given call to `llama_decode` presumably you would never want both. For the `gritlm` example, I actually just made two contexts, one for generation one for embeddings. Another option would be to add a `llama_set_embeddings` function.